### ISSUE TITLE
关于新增模型遇到的问题

### ISSUE THREAD

**Issue by lmk123568** [PaddlePaddle/PaddleDetection#3022]
需要基于paddle开发新的检测模型，新增了shufflenetv2魔改的模型`shufflenet_v2.py`
```
import paddle
import paddle.nn as nn
from ppdet.core.workspace import register, serializable
from ..shape_spec import ShapeSpec

__all__ = ['ShuffleNetV2']

class ShuffleV2Block(nn.Layer):
    def __init__(self, inp, oup, mid_channels, *, ksize, stride, r=16):
        super(ShuffleV2Block, self).__init__()
        self.stride = stride
        assert stride in [1, 2]

        self.mid_channels = mid_channels
        self.ksize = ksize
        pad = ksize // 2
        self.pad = pad
        self.inp = inp
        
        outputs = oup - inp  

        branch_main = [
            # pw
            nn.Conv2D(inp, mid_channels, 1, 1, 0, bias_attr=True),
            nn.BatchNorm2D(mid_channels),
            nn.ReLU(),
            # dw
            nn.Conv2D(mid_channels, mid_channels, ksize, stride, pad, groups=mid_channels, bias_attr=True),
            nn.BatchNorm2D(mid_channels),
            # pw-linear
            nn.Conv2D(mid_channels, outputs, 1, 1, 0, bias_attr=True),
            nn.BatchNorm2D(outputs),
            nn.ReLU(),
        ]
        self.branch_main = nn.Sequential(*branch_main)

        if stride == 2:
            branch_proj = [
                # dw
                nn.Conv2D(inp, inp, ksize, stride, pad, groups=inp, bias_attr=True),
                nn.BatchNorm2D(inp),
                # pw-linear
                nn.Conv2D(inp, inp, 1, 1, 0, bias_attr=True),
                nn.BatchNorm2D(inp),
                nn.ReLU(),
            ]
            self.branch_proj = nn.Sequential(*branch_proj)
        else:
            self.branch_proj = None

    def forward(self, old_x):
        if self.stride==1:
            x_proj, x = self.channel_shuffle(old_x)
            return paddle.concat((x_proj, self.branch_main(x)), 1)
        elif self.stride==2:
            x_proj = old_x
            x = old_x
            return paddle.concat((self.branch_proj(x_proj), self.branch_main(x)), 1)

    def channel_shuffle(self, x):
        batchsize, num_channels, height, width = x.shape
        assert (num_channels % 4 == 0)
        x = x.reshape([batchsize * num_channels // 2, 2, height * width])
        x = x.transpose([1, 0, 2])
        x = x.reshape([2, -1, num_channels // 2, height, width])
        return x[0], x[1]


@register
@serializable
class ShuffleNetV2(nn.Layer):
    def __init__(self):
        """
        ShuffleNet, see https://

        Args:
            freeze_at (int): freeze the backbone at which stage
            return_idx (list): index of stages whose feature maps are returned
        """
        super(ShuffleNetV2, self).__init__()

        self.stage_repeats = [4, 8, 4]
        self.r = 16
        self.stage_out_channels = [-1, 24, 256, 512, 1024]


        # building first layer
        input_channel = self.stage_out_channels[1]
        self.stage0 = nn.Sequential(
            nn.Conv2D(3, input_channel, 3, 2, 1, bias_attr=True),
            nn.BatchNorm2D(input_channel),
            nn.ReLU(),
        )

        self.maxpool = nn.MaxPool2D(kernel_size=3, stride=2, padding=1)

        
        for idxstage in range(len(self.stage_repeats)):
            numrepeat = self.stage_repeats[idxstage]
            output_channel = self.stage_out_channels[idxstage+2]

            layers = []
            for i in range(numrepeat):
                if i == 0:
                    layers.append(ShuffleV2Block(input_channel, output_channel,
                                                mid_channels=output_channel // 2, ksize=3, stride=2, r=self.r))
                else:
                    layers.append(ShuffleV2Block(input_channel // 2, output_channel,
                                                mid_channels=output_channel // 2, ksize=3, stride=1, r=self.r))

                input_channel = output_channel

            if idxstage == 0:
                self.stage1= nn.Sequential(*layers)
            if idxstage == 1:
                self.stage2= nn.Sequential(*layers)
            if idxstage == 2:
                self.stage3= nn.Sequential(*layers)            


        self._initialize_weights()


    def _initialize_weights(self):
        normal = nn.initializer.KaimingNormal()
        constant = nn.initializer.Constant(value=0)
        for name, m in self.named_sublayers():
            if isinstance(m, nn.Conv2D):
                normal(m.weight)
                if m.bias is not None:
                    constant(m.bias)
            elif isinstance(m, nn.Linear):
                normal(m.weight)
                if m.bias is not None:
                    constant(m.bias)


    def forward(self, x):

        blocks = []


        x = self.stage0(inputs['image'])
        x = self.maxpool(x)
        x = self.stage1(x)
        blocks.append(x)
        x = self.stage2(x)
        blocks.append(x)
        x = self.stage3(x)
        blocks.append(x)


        return blocks


    @property
    def out_shape(self):
        return [ShapeSpec(channels=c) for c in [256,512,1024]]
```
在backbone文件下`__init__`增加
```
from . import shufflenet_v2
from .shufflenet_v2 import *
```
然后在YOLOv3配置下`__base__`新增文件`yolov3_shufflenet_v2`
```
architecture: YOLOv3
pretrain_weights: https://paddledet.bj.bcebos.com/models/pretrained/MobileNetV1_pretrained.pdparams
norm_type: sync_bn

YOLOv3:
  backbone: ShuffleNetV2
  neck: YOLOv3FPN
  yolo_head: YOLOv3Head
  post_process: BBoxPostProcess


# use default config
# YOLOv3FPN:

YOLOv3Head:
  anchors: [[10, 13], [16, 30], [33, 23],
            [30, 61], [62, 45], [59, 119],
            [116, 90], [156, 198], [373, 326]]
  anchor_masks: [[6, 7, 8], [3, 4, 5], [0, 1, 2]]
  loss: YOLOv3Loss

YOLOv3Loss:
  ignore_thresh: 0.7
  downsample: [32, 16, 8]
  label_smooth: false

BBoxPostProcess:
  decode:
    name: YOLOBox
    conf_thresh: 0.005
    downsample_ratio: 32
    clip_bbox: true
  nms:
    name: MultiClassNMS
    keep_top_k: 100
    score_threshold: 0.01
    nms_threshold: 0.45
    nms_top_k: 1000
```
运行`yolov3_shufflenet_v2_roadsign.yaml`
```
_BASE_: [
  '../datasets/roadsign_voc.yml',
  '../runtime.yml',
  '_base_/yolov3_shufflenet_v2.yml',
  '_base_/yolov3_reader.yml',
]
pretrain_weights: None
weights: output/yolov3_mobilenet_v1_roadsign/model_final

YOLOv3Loss:
  ignore_thresh: 0.7
  label_smooth: true

snapshot_epoch: 2
epoch: 40

LearningRate:
  base_lr: 0.0001
  schedulers:
  - !PiecewiseDecay
    gamma: 0.1
    milestones: [32, 36]
  - !LinearWarmup
    start_factor: 0.3333333333333333
    steps: 100

OptimizerBuilder:
  optimizer:
    momentum: 0.9
    type: Momentum
  regularizer:
    factor: 0.0005
    type: L2

```
报错
![45414bb4d2b9ef7e76242d6e0b5b5ee](https://user-images.githubusercontent.com/60082702/118365828-dc5f7300-b5d0-11eb-8409-a4a4868257c5.png)

**Comment by jerrywgz** [comment]
我这里用你的代码验证了下，是register成功的，你的paddledet是否通过pip的方式安装了，可以pip uninstall paddledet 然后再运行训练代码

**Comment by lmk123568** [comment]
> 我这里用你的代码验证了下，是register成功的，你的paddledet是否通过pip的方式安装了，可以pip uninstall paddledet 然后再运行训练代码

感谢回复，是通过pip安装paddledet，没懂后半段，是uninstall之后再install paddledet吗

**Comment by lmk123568** [comment]
> 我这里用你的代码验证了下，是register成功的，你的paddledet是否通过pip的方式安装了，可以pip uninstall paddledet 然后再运行训练代码

试了很多遍，还是不行

**Comment by jerrywgz** [comment]
可以先pip uninstall paddledet， 然后pip install -r requirement.txt 之后直接启动训练

**Comment by lmk123568** [comment]
谢谢，已解决，不过我还是没明白pip install paddledet，和pip install -r requirement.txt区别，为什么这样做就能运行了，以后如果基于paddledet开发新模型，该选择pip安装还是源码编译安装

**Comment by lmk123568** [comment]
> 可以先pip uninstall paddledet， 然后pip install -r requirement.txt 之后直接启动训练

谢谢，已解决，不过我还是没明白pip install paddledet，和pip install -r requirement.txt区别，为什么这样做就能运行了，以后如果基于paddledet开发新模型，该选择pip安装还是源码编译安装

**Comment by jerrywgz** [comment]
如果需要基于源码二次开发，建议使用源码安装

**Comment by lmk123568** [comment]
> 如果需要基于源码二次开发，建议使用源码安装
源码安装之后，修改代码后，还需要再次安装吗，或者运行pip setup.py develop

**Comment by lmk123568** [comment]
> 如果需要基于源码二次开发，建议使用源码安装

源码安装之后，修改代码后，还需要再次安装吗，或者运行pip setup.py develop
